File: C/11-container-with-most-water/main.go
package main

import (
	"fmt"
)

func maxArea(height []int) int {
	start, end := 0, len(height)-1
	max := 0

	for start < end {
		base := start
		isStart := true
		if height[base] > height[end] {
			base = end
			isStart = false
		}
		fmt.Println(base, start, end)
		if height[base]*(end-start) > max {
			max = height[base] * (end - start)
		}

		if isStart {
			start++
		} else {
			end--
		}
	}
	return max
}

func main() {
	fmt.Println(maxArea([]int{1, 8, 6, 2, 5, 4, 8, 3, 7}))
}

File: C/11-container-with-most-water/main.rs
impl Solution {
  pub fn max_area(height: Vec<i32>) -> i32 {
    let mut s: usize = 0;
    let mut e: usize = height.len() - 1;

    let mut ans: i32 = 0;
    while s < e {
      ans = ans.max(height[s].min(height[e]) * (e - s) as i32);
      if height[s] < height[e] {
        s += 1;
      } else {
        e -= 1;
      }
    }
    ans
  }
}
